### PR TITLE
Add `Init` method to `TagHelper`s.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- The init method allows multiple `TagHelper`s to inject data into the `context.Items` bag to properly function when running in unison with other `TagHelper`s that need to communicate with children.
- Transition `GetChildContentAsync` from `TagHelperContext` to `TagHelperOutput`.
- Move `TagHelperContext.GetChildContentAsync` tests to `TagHelperOutputTest`.
- Added `Init` test to ensure `TagHelperRunner` calls it in the correct order.

#571

### CODE REVIEW

**Comment by NTaylorMullen** [comment]
Updated: Moved `GetChildContentAsync`.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelper.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelper.cs]: Why is this needed?

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs [hunk 2]
```diff
@@ -13,8 +13,6 @@ namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
     /// </summary>
     public class TagHelperContext
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs]: Does this class still need all the `using`s e.g. for `System.Threading.Tasks`?

**Comment by dougbu** [comment]
This looks generally good. Ready for non-design and the MVC reaction PR.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs]: done

**Comment by NTaylorMullen** [comment]
Updated to not be design. Working on MVC react.

**Comment by NTaylorMullen** [comment]
MVC Reaction: https://github.com/aspnet/Mvc/pull/3349

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs [hunk 3]
```diff
@@ -17,6 +17,16 @@ public interface ITagHelper
         int Order { get; }
 
         /// <summary>
+        /// Initializes the <see cref="ITagHelper"/> with the given <paramref name="context"/>.
+        /// </summary>
+        /// <param name="context">Contains information associated with the current HTML tag.</param>
+        /// <remarks>Any additions to <see cref="TagHelperContext.Items"/> should be done within this method to ensure
+        /// they're added prior to executing the children content. When more than one <see cref="ITagHelper"/> runs
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs]: Pick one: `executing the children` or `determining the child content`. "children content" is a mangle.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs [hunk 4]
```diff
@@ -17,6 +17,16 @@ public interface ITagHelper
...
+        /// on the same element <see cref="TagHelperOutput.GetChildContentAsync"/> may be invoked prior to
+        /// <see cref="ProcessAsync(TagHelperContext, TagHelperOutput)"/> being called.</remarks>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs]: remove `being called`

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 5]
```diff
@@ -21,17 +24,32 @@ internal TagHelperOutput(string tagName)
...
         /// <param name="attributes">The HTML attributes.</param>
+        public TagHelperOutput(string tagName, TagHelperAttributeList attributes)
+            : this(
+                tagName,
+                attributes,
+                (cachedResult) => Task.FromResult<TagHelperContent>(new DefaultTagHelperContent()))
+        {
+        }
+
         public TagHelperOutput(
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: This method needs docs.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: :+1:

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 6]
```diff
@@ -21,17 +24,32 @@ internal TagHelperOutput(string tagName)
         /// </summary>
         /// <param name="tagName">The HTML element's tag name.</param>
         /// <param name="attributes">The HTML attributes.</param>
+        public TagHelperOutput(string tagName, TagHelperAttributeList attributes)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Can we remove this constructor? If not at least make it `internal` (for testing).

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs [hunk 7]
```diff
@@ -27,8 +24,7 @@ public class TagHelperContext
         public TagHelperContext(
             IEnumerable<IReadOnlyTagHelperAttribute> allAttributes,
             IDictionary<object, object> items,
-            string uniqueId,
-            Func<bool, Task<TagHelperContent>> getChildContentAsync)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs]: Remove the `<param>` for this parameter.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs [hunk 8]
```diff
@@ -17,6 +17,16 @@ public interface ITagHelper
         int Order { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs]: "executed" is now vague. Mention `Init()` and `Process()` / `ProcessAsync()` methods are both invoked in given order and `Init()` methods are all done prior to `Process()` / `ProcessAsync()` methods.

**Comment by dougbu** [comment]
:watch:

**Comment by Eilon** [comment]
Looks nice.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 6 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Ya, I'll nuke it.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs (hunk 8 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs]: done.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs (hunk 7 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs]: done

**Comment by NTaylorMullen** [comment]
Updated.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs (hunk 8 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ITagHelper.cs]: Unfortunately `<remarks>` don't show up in IntelliSense and @Eilon loves IntelliSense. Suggest using this instead of the vaguer "execution order" in the `<summary>`.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 9]
```diff
@@ -115,5 +130,26 @@ public void SuppressOutput()
             PostContent.Clear();
             PostElement.Clear();
         }
+
+        /// <summary>
+        /// A delegate used to execute and retrieve the rendered child content asynchronously.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> that when executed returns content rendered by children.</returns>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: We usually say "on completion" rather than "when executed".

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 10]
```diff
@@ -115,5 +130,26 @@ public void SuppressOutput()
             PostContent.Clear();
             PostElement.Clear();
         }
+
+        /// <summary>
+        /// A delegate used to execute and retrieve the rendered child content asynchronously.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: "execute _children_" might make this read a bit better. Odd to think of executing child content. Applies in the associated constructor `<param>` and the other overload.

**Comment by dougbu** [comment]
:shipit: with a couple of small fixes
